docs/docs/reference/release-notes/migrating-from-v2-to-v3.md

```diff
@@ -0,0 +1,588 @@
+  }
+
+https://www.gatsbyjs.com/docs/actions/#createTypes
+```
+
+If this warning is displayed for a type defined by some plugin - open an issue in the plugin repo
+with suggestion to upgrade (and a link to this guide).
+
+You can still fix those warnings temporarily in your site's `gatsby-node.js` until it is fixed in the plugin.
+
+If you don't see any warnings - you are safe to upgrade to v3. Read more about custom extensions in [this blog post](https://www.gatsbyjs.com/blog/2019-05-17-improvements-to-schema-customization/).
+
+### Schema Customization: Removed `noDefaultResolvers` argument from inference directives
+
+Search for `noDefaultResolvers` entries and remove them:
+
+```diff:title=gatsby-node.js
+exports.createSchemaCustomization = ({ actions }) => {
+  const { createTypes } = actions
+  createTypes(`
+-   type Foo implements Node @infer(noDefaultResolvers: true)
++   type Foo implements Node @infer
+    {
+      id: ID!
+    }
+
+-   type Bar implements Node @dontInfer(noDefaultResolvers: true)
++   type Foo implements Node @dontInfer
+    {
+      id: ID!
+    }
+
+  `)
+}
+```
+
+[Read deprecation announcement](https://www.gatsbyjs.com/blog/2019-05-17-improvements-to-schema-customization/#-nodefaultresolvers-and-inference-modes).
+
+### Schema Customization: Remove `many` argument from `childOf` directive
+
+It is no longer needed in Gatsby v3:
+
+```diff:title=gatsby-node.js
+exports.createSchemaCustomization = ({ actions }) => {
+  const { createTypes } = actions
+  createTypes(`
+-   type Foo implements Node @childOf(types: ["Bar"], many: true)
++   type Foo implements Node @childOf(types: ["Bar"])
+    {
+      id: ID!
+    }
+  `)
+}
+```
+
+### Schema Customization: Consistent return for `nodeModel.runQuery`
+
+For Gatsby v2, `nodeModel.runQuery` with `firstOnly: false` returns `null` when nothing is found.
+In v3 it returns an empty array instead.
+
+To upgrade, find all occurrences of `runQuery` (with `firstOnly: false` or not set) and make sure checks for emptiness
+are correct:
+
+```diff:title=gatsby-node.js
+exports.createResolvers = ({ createResolvers }) => {
+  const resolvers = {
+    Foo: {
+      bars: {
+        resolve(source, args, context, info) {
+          const result = context.nodeModel.runQuery({
+            query: {/* */},
+            type: "Bar",
+            firstOnly: false,
+          })
+-         if (result === null) {
++         if (result.length === 0) {
+            throw new Error("Not found!")
+          }
+          return result
+        },
+      },
+    },
+  }
+  createResolvers(resolvers)
+}
+```
+
+**Note:** when using argument `firstOnly: true` the returned value is `object` or `null`.
+So do not confuse those two cases.
+
+## Resolving Deprecations
+
+This section explains deprecations that were made for Gatsby v3. The old behaviors will be removed in v4. You can still use the old behaviors in v3 but we recommend updating to the new signatures.
```

Comment:
This section confuses me. All of these are mentioned above as changes users should make. Is the point that the behaviors still work but are merely deprecated? It seems like they were deprecated in v2 and are now removed?

Reply:
In v1 it was this behavior: `deleteNode(node.id)`
In v2 it was this behavior: `deleteNode({ node })`
Now in v3 it's: `deleteNode(node)`

So the breaking changes are the v2 deprecation, and the deprecations are the old v2 behavior.

Reply:
Ah, it's nuanced enough that I missed it. Might it make sense to consolidate each of these with their previous sections? So that we mention the current behavior and then point out what will be deprecated in the future? Not sold on that approach, but figured I'd mention.

Reply:
The thing is, if we mix breaking changes & deprecations, then it'll get even more confusing I think.

Reply:
```suggestion
This section explains deprecations that were made for Gatsby v3. These old behaviors will be removed in v4, at which point they will no longer work. For now, you can still use the old behaviors in v3, but we recommend updating to the new signatures to make future updates easier.
```